// @editor/user/source/utils/network.ts
'use strict';

const { parse } = require('url');
const { stringify } = require('querystring');
const { request: httpsRequest } = require('https');

/**
 * 发送 post 消息到服务器
 * @param {string} url 
 * @param {object} data 
 */
export function sendPostRequest(url: string, data: any) {
	return new Promise((resolve, reject) => {
		let urlItem = parse(url);
		data = stringify(data || {});
	
		let headers = {
			'Content-Type': 'application/x-www-form-urlencoded',
			'Content-Length': data.length,
			'User-Agent': 'Mozilla/5.0 (Linux; Android 5.0; SM-G900P Build/LRX21T) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/48.0.2564.23 Mobile Safari/537.36'
		};
	
		let sendData = {
			method: 'POST',
			host: urlItem.hostname,
			port: urlItem.port || 443,
			path: urlItem.pathname,
			headers: headers,
			ciphers: 'ALL', // 需要加这两个参数，解决问题 throwing EPROTO error
            secureProtocol: 'TLSv1_method',
		};
	
		let body = '';
		let request = httpsRequest(sendData, (response: any) => {
			if (response.statusCode !== 200) {
				reject(new Error('Connect Failed'));
				return;
			}
	
			response.on('data', (buf: any) => {
				body += buf;
			}).on('end', () => {
				var json = null;
				try {
					json = JSON.parse(body);
				} catch (error) {
					reject(error);
					return;
				}
				resolve(json);
			});
		}).on('error', (error: Error) => {
			reject(error);
		}).setTimeout(8000, () => {
			reject(new Error('timeout'));
		});
	
		request.write(data);
		request.end();
	});
};

/**
 * 发送 get 消息到服务器
 * @param {string} url 
 * @param {object} data 
 */
export function sendGetRequest(url: string, data: any) {
	return new Promise((resolve, reject) => {
		var urlItem = parse(url);
		data = stringify(data || {});

		var headers = {
			'User-Agent': 'Mozilla/5.0 (Linux; Android 5.0; SM-G900P Build/LRX21T) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/48.0.2564.23 Mobile Safari/537.36'
		};

		var sendData = {
			method: 'GET',
			host: urlItem.hostname,
			port: urlItem.port || 443,
			path: urlItem.pathname + '?' + data,
			headers: headers,
			ciphers: 'ALL', // 需要加这两个参数，解决问题 throwing EPROTO error
			secureProtocol: 'TLSv1_method',
		};

		var body = '';
		var request = httpsRequest(sendData, (response: any) => {
			if (response.statusCode !== 200) {
				reject(new Error('Connect Failed'));
				return;
			}

			response.on('data', (buf: any) => {
				body += buf;
			}).on('end', () => {
				var json = null;
				try {
					json = JSON.parse(body);
				} catch (error) {
					return reject(error);
				}
				resolve(json);
			});
		}).on('error', (error: Error) => {
			reject(error);
		}).setTimeout(8000, () => {
			reject(new Error('timeout'));
		});

		request.end();
	});
};
