// @editor/user/source/index.ts
'use strict';

const { EventEmitter } = require('events');
const { network } = require('./utils');
import { ChildProcess } from 'child_process';

interface UserInfo {
    session_id: string;
    session_key: string;
    cocos_uid: string;
    email: string;
    nickname: string;
}

// user 登陆后服务器反馈的验证数据
const info: UserInfo = {
    session_id: '',
    session_key: '',
    cocos_uid: '',
    email: '',
    nickname: '',
};

const EventManager = new EventEmitter();

/**
 * 设置 info 数据
 * @param {object} data
 */
export function setData(data: UserInfo) {
    info.session_id = data.session_id;
    info.session_key = data.session_key;
    info.cocos_uid = data.cocos_uid;
    info.email = data.email;
    info.nickname = data.nickname;
    EventManager.emit('info-update');
};

/**
 * 读取 info 数据
 */
export function getData () {
    return info;
}

/**
 * 检查是否已经登陆
 */
export async function isLoggedIn () {
    // 没有记录数据，则未登陆
    if (!info.session_key) {
        return false;
    }

    // 请求服务器数据
    let url = 'https://creator-api.cocos.com/api/account/is_online';
    let result;

    result = await network.sendPostRequest(url, {
        session_key: info.session_key
    });

    if (result.status !== 0) {
        // 状态码不为 0, 则接口识别不正确
        return false;
    }

    // 刷新缓存数据，必须刷新，没次请求后，session 都会更改
    info.session_id = result.data.session_id;
    info.session_key = result.data.session_key;
    info.cocos_uid = result.data.cocos_uid;
    info.email = result.data.email;
    info.nickname = result.data.nickname;
    EventManager.emit('info-update');

    return true;
}

/**
 * 用户登陆
 * 登陆失败会直接抛出异常
 * @param {string} username 
 * @param {string} password 
 */
export async function login(username: string, password: string) {
    let url = 'https://creator-api.cocos.com/api/account/signin';
    let result;

    result = await network.sendPostRequest(url, {
        username, password,
    });

    if (result.status !== 0) {
        throw new Error(result.status + ': ' + result.msg);
    }

    // 刷新缓存数据，必须刷新，没次请求后，session 都会更改
    info.session_id = result.data.session_id;
    info.session_key = result.data.session_key;
    info.cocos_uid = result.data.cocos_uid;
    info.email = result.data.email;
    info.nickname = result.data.nickname;
    EventManager.emit('info-update');

    // 发送登陆事件
    EventManager.emit('login');

    return {
        nickname: info.nickname,
        email: info.email,
    };
}

/**
 * 退出已经登陆的用户
 * 如果退出失败，直接抛出异常
 */
export async function logout() {
    let url = 'https://creator-api.cocos.com/api/user/signout';
    let result;

    result = await network.sendPostRequest(url, {
        session_id: info.session_id,
    });

    if (result.status !== 0) {
        throw new Error(result.status + ': ' + result.msg);
    }

    info.session_id = '';
    info.session_key = '';
    info.cocos_uid = '';
    info.email = '';
    info.nickname = '';
    EventManager.emit('info-update');

    // 发送登出事件
    EventManager.emit('logout');

    return true;
}

/**
 * 获取用户的 user token
 */
export async function getUserToken() {
    let url = 'https://creator-api.cocos.com/api/user/cocos_token';
    let result;

    result = await network.sendPostRequest(url, {
        session_id: info.session_id,
    });

    if (result.status !== 0) {
        throw new Error(result.status + ': ' + result.msg);
    }

    return result.data;
}

/**
 * 获取插件的 session code
 * @param {*} pluginId 
 */
export async function getSessionCode(pluginId: number) {
    let url = 'https://creator-api.cocos.com/api/session/code';
    let result;

    result = network.sendPostRequest(url, {
        session_id: info.session_id,
        plugin_id: pluginId
    });

    if (result.status !== 0) {
        throw new Error(result.status + ': ' + result.msg);
    }

    return result.data;
}

/**
 * 绑定用户事件
 * @param args 
 */
export function on(...args: any[]) {
    EventManager.on(...args);
}

/**
 * 绑定单次用户事件
 * @param args 
 */
export function once(...args: any[]) {
    EventManager.once(...args);
}

/**
 * 解除绑定的事件
 * @param args 
 */
export function removeListener(...args: any[]) {
    EventManager.removeListener(...args);
}

////////// 支持子进程 //////////

const children: ChildProcess[] = [];

/**
 * 添加一个受管理的子进程
 * @param child 
 */
export function addChildProcess(child: ChildProcess) {
    if (children.indexOf(child) !== -1) {
        return;
    }
    children.push(child);

    // @ts-ignore
    child.__handle = async function(options: any) {
        // 如果不是我们监听的消息，则忽略
        if (!options || !options.channel || options.channel !== 'cocos-user') {
            return;
        }

        const func = module.exports[options.func];
        if (!func) {
            return;
        }

        let result: any;

        try {
            result = await func(...(options.params || []));
        } catch(error) {
            child.send({
                channel: 'cocos-user',
                id: options.id,
                error: error.message,
                result: null,
            });
            return;
        }

        child.send({
            channel: 'cocos-user',
            id: options.id,
            error: null,
            result: result,
        });
    }

    // @ts-ignore
    child.on('message', child.__handle);
}

/**
 * 移除一个受管理的子进程
 * @param child 
 */
export function removeChildProcess(child: ChildProcess) {
    const index = children.indexOf(child);
    if (index === -1) {
        return;
    }
    children.splice(index, 1);
    // @ts-ignore
    child.removeListener('message', child.__handle);
}

// 转发 login 事件给每个子进程
EventManager.on('login', () => {
    children.forEach((child) => {
        child.send({
            channel: 'cocos-user',
            emit: 'login',
            result: [],

        });
    });
});

// 转发 logout 事件给每个子进程
EventManager.on('logout', () => {
    children.forEach((child) => {
        child.send({
            channel: 'cocos-user',
            emit: 'logout',
            result: [],

        });
    });
});

// 转发 info-update 事件给每个子进程
EventManager.on('info-update', () => {
    children.forEach((child) => {
        child.send({
            channel: 'cocos-user',
            emit: 'info-update',
            result: [],

        });
    });
});

////////// 支持 electron 渲染进程 //////////

// @ts-ignore
if (process.type === 'browser') {
    const ipc = require('@base/electron-base-ipc');
    ipc.on('cocos-user', async (event: any, name: string, args: any[]) => {
        const func = module.exports[name];
        if (!func) {
            return;
        }
    
        let result: any;
    
        try {
            result = await func(...(args || []));
        } catch(error) {
            event.reply(error, null);
            return;
        }
        event.reply(null, result);
    });

    // 转发 login 事件给每个子进程
    EventManager.on('login', () => {
        ipc.broadcast('cocos-user-emit', 'login');
    });

    // 转发 logout 事件给每个渲染进程
    EventManager.on('logout', () => {
        ipc.broadcast('cocos-user-emit', 'logout');
    });

    // 转发 info-update 事件给每个渲染进程
    EventManager.on('info-update', () => {
        ipc.broadcast('cocos-user-emit', 'info-update');
    });
}
