// @editor/user/source/platform/child.ts
'use strict';

const { EventEmitter } = require('events');

/**
 * 子进程数据，通过 process 和主进程进行数据交互
 */

interface sendCache {
    resolve: Function;
    reject: Function;
}

const EventManager = new EventEmitter();

let sendID = 1;
const sendMap: { [index: number]: sendCache } = {};

// 接收主进程发过来的消息
process.on('message', (options: any) => {
    // 如果不是我们监听的消息，则忽略
    if (!options || !options.channel || options.channel !== 'cocos-user') {
        return;
    }

    // 主进程发送过来的事件触发
    if (options.emit) {
        emit(options.emit, ...(options.result || []));
        return;
    }

    // 检查等待返回的队列
    const sendCache = sendMap[options.id];
    if (!sendCache) {
        return;
    }
    // 删除等待兑现
    delete sendMap[options.id];
    // 如果有错误的情况，触发 reject
    if (options.error) {
        return sendCache.reject(new Error(options.error));
    }
    // 否则正常返回数据
    return sendCache.resolve(options.result);
});

/**
 * 获取用户数据
 */
export async function getData() {
    return new Promise((resolve, reject) => {
        // 缓存
        sendMap[sendID] = { resolve, reject };
        process.send && process.send({
            channel: 'cocos-user',
            id: sendID++,
            func: 'getData',
            params: [],
        });
    });
}

/**
 * 检查用户是否登陆
 */
export async function isLoggedIn() {
    return new Promise((resolve, reject) => {
        // 缓存
        sendMap[sendID] = { resolve, reject };
        process.send && process.send({
            channel: 'cocos-user',
            id: sendID++,
            func: 'isLoggedIn',
            params: [],
        });
    });
}

/**
 * 用户登陆
 * @param username 
 * @param password 
 */
export function login(username: string, password: string) {
    return new Promise((resolve, reject) => {
        // 缓存
        sendMap[sendID] = { resolve, reject };
        process.send && process.send({
            channel: 'cocos-user',
            id: sendID++,
            func: 'login',
            params: [username, password],
        });
    });
}

/**
 * 用户退出
 */
export function logout() {
    return new Promise((resolve, reject) => {
        // 缓存
        sendMap[sendID] = { resolve, reject };
        process.send && process.send({
            channel: 'cocos-user',
            id: sendID++,
            func: 'logout',
            params: [],
        });
    });
}

/**
 * 获取用户的 token
 */
export function getUserToken() {
    return new Promise((resolve, reject) => {
        // 缓存
        sendMap[sendID] = { resolve, reject };
        process.send && process.send({
            channel: 'cocos-user',
            id: sendID++,
            func: 'getUserToken',
            params: [],
        });
    });
}

/**
 * 获取一个插件的 session code
 * @param pluginId 
 */
export function getSessionCode(pluginId: number) {
    return new Promise((resolve, reject) => {
        // 缓存
        sendMap[sendID] = { resolve, reject };
        process.send && process.send({
            channel: 'cocos-user',
            id: sendID++,
            func: 'getSessionCode',
            params: [pluginId],
        });
    });
}

/**
 * 触发绑定的事件
 * @param args 
 */
export function emit(...args: any[]) {
    EventManager.emit(...args);
}

/**
 * 绑定用户事件
 * @param args 
 */
export function on(...args: any[]) {
    EventManager.on(...args);
}

/**
 * 绑定单次用户事件
 * @param args 
 */
export function once(...args: any[]) {
    EventManager.once(...args);
}

/**
 * 解除绑定的事件
 * @param args 
 */
export function removeListener(...args: any[]) {
    EventManager.removeListener(...args);
}
