// @editor/user/source/platform/renderer.ts
'use strict';

const { EventEmitter } = require('events');
const ipc = require('@base/electron-base-ipc');

/**
 * Electron 渲染进程
 */

const EventManager = new EventEmitter();

ipc.on('cocos-user-emit', (event: any, action: string, ...args: any[]) => {
    emit(action, ...args);
});

export async function getData() {
    return new Promise((resolve, reject) => {
        ipc
            .send('cocos-user', 'getData', [])
            .callback((error: Error, result: any) => {
                if (error) {
                    return reject(error);
                }
                resolve(result);
            });
    });
}

export async function isLoggedIn() {
    return new Promise((resolve, reject) => {
        ipc
            .send('cocos-user', 'isLoggedIn', [])
            .callback((error: Error, result: any) => {
                if (error) {
                    return reject(error);
                }
                resolve(result);
            });
    });
}

export function login(username: string, password: string) {
    return new Promise((resolve, reject) => {
        ipc
            .send('cocos-user', 'login', [username, password])
            .callback((error: Error, result: any) => {
                if (error) {
                    return reject(error);
                }
                resolve(result);
            });
    });
}

export function logout() {
    return new Promise((resolve, reject) => {
        ipc
            .send('cocos-user', 'logout', [])
            .callback((error: Error, result: any) => {
                if (error) {
                    return reject(error);
                }
                resolve(result);
            });
    });
}

export function getUserToken() {
    return new Promise((resolve, reject) => {
        ipc
            .send('cocos-user', 'getUserToken', [])
            .callback((error: Error, result: any) => {
                if (error) {
                    return reject(error);
                }
                resolve(result);
            });
    });
}

export function getSessionCode(pluginId: number) {
    return new Promise((resolve, reject) => {
        ipc
            .send('cocos-user', 'getSessionCode', [pluginId])
            .callback((error: Error, result: any) => {
                if (error) {
                    return reject(error);
                }
                resolve(result);
            });
    });
}

/**
 * 触发绑定的事件
 * @param args 
 */
export function emit(...args: any[]) {
    EventManager.emit(...args);
}

/**
 * 绑定用户事件
 * @param args 
 */
export function on(...args: any[]) {
    EventManager.on(...args);
}

/**
 * 绑定单次用户事件
 * @param args 
 */
export function once(...args: any[]) {
    EventManager.once(...args);
}

/**
 * 解除绑定的事件
 * @param args 
 */
export function removeListener(...args: any[]) {
    EventManager.removeListener(...args);
}
